cairo/ethereum/cancun/trie.cairo

```diff
@@ -329,24 +329,67 @@ func encode_node{range_check_ptr, bitwise_ptr: BitwiseBuiltin*, keccak_ptr: Kecc
     return encoded;
 }
 
-// func copy_trie(trie: Trie[K, V]) -> Trie[K, V] {
-//     // Implementation:
-//     // return Trie(trie.secured, trie.default, copy.copy(trie._data))
-// }
+// @notice Copies the trie to a new segment.
+// @dev This function simply creates a new segment for the new dict and associates it with the
+// dict_tracker of the source dict.
+func copy_trieAddressAccount{range_check_ptr, trie: TrieAddressAccount}() -> TrieAddressAccount {
+    alloc_locals;
+    // TODO: soundness
+    // We need to ensure it is sound when finalizing that copy.
+    // The full design is:
+    // - We create a new segment for the new dict
+    // - We copy the python dict tracker and associate it with that new segment
+    // - When interacting with the copied trie, we use the new segment with the new dict_ptr
+    // - If the state reverts, then upon squashing that copy, we:
+    //  - copy all the prev_keys in the new segment to the main segment (as if they were read from the new segment)
+    //  - delete the new segment
+    //  - This ensures that when squashing the main segment, we ensure that the data read in the new segment matched the data from the main segment.
```

Comment:
Shouldn't the corresponding squashing function be done in this PR and be tested in revert and valid case?
This is a critical part of the trie copy behavior

Reply:
https://github.com/kkrt-labs/keth/issues/389